bigframes/core/io.py

```diff
@@ -16,11 +16,43 @@
 
 import datetime
 import textwrap
-from typing import Dict, Union
+from typing import Dict, Optional, Sequence, Tuple, Union
 
 import google.cloud.bigquery as bigquery
 
 IO_ORDERING_ID = "bqdf_row_nums"
+MAX_LABELS_COUNT = 64
+
+
+def create_job_configs_labels(
+    job_config: Optional[bigquery.job.QueryJobConfig],
+    api_methods: Sequence[str],
+) -> Tuple[Dict[str, str], bigquery.job.QueryJobConfig]:
+    if job_config is not None:
+        # If there is no label set
+        if job_config.labels is None:
+            label_values = api_methods
+        else:
+            cur_labels: Sequence[str] = [*job_config.labels.values()]
+            cur_labels_len = len(job_config.labels)
+            api_methods_len = len(api_methods)
+            # If the total number of labels is under the limit of labels count
+            if cur_labels_len + api_methods_len <= MAX_LABELS_COUNT:
+                label_values = list(cur_labels) + list(api_methods)
+            # We capture the latest label if it is out of the length limit of labels count
+            else:
+                added_api_len = cur_labels_len + api_methods_len - MAX_LABELS_COUNT
+                label_values = list(cur_labels) + list(api_methods)[-added_api_len:]
+    else:
+        job_config = bigquery.QueryJobConfig()
+        label_values = api_methods
+
+    labels = {}
+    for i, label_value in enumerate(label_values):
+        label_key = "bigframes-api-" + str(i)
+        labels[label_key] = label_value
+    job_config.labels = labels
```

Comment:
Mutating the job config will make testing more difficult.

Reply:
Pass existing labels as dictionary and return dictionary now. We test if the the output dictionary is expected.